### PR TITLE
chore: start using plugin-framework package

### PR DESCRIPTION
**Pull request by priteshbandi**
chore: start using plugin-framework package.

**Note:** we introduced a small breaking change; the type for `GenerateSignatureResponse#SigningAlgorithm` has been changed from `string` to `plugin.SignatureAlgorithm`

### CODE REVIEW

File: plugin/proto/metadata.go
Comment (Two-Hearts) [inline plugin/proto/metadata.go:39]: Instead of changing this function (which is a breaking change), we should move this function to `notation-plugin-framework-go/plugin/metadata.go`.

File: signer/plugin.go
Comment (Two-Hearts) [inline signer/plugin.go:82]: With the above change, we won't need this update.

File: plugin/plugin.go
Comment (shizhMSFT) [inline plugin/plugin.go:37]: To mark an API deprecated, you need to update the documentation by adding `deprecated: <reason_and_recommendation>`.

See https://go.dev/wiki/Deprecated for more details.

File: plugin/proto/errors.go
Comment (shizhMSFT) [inline plugin/proto/errors.go:49]: Since `jsonErr` is not exported, you can remove line 49 and use `plugin.Error` directly in the code.

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: go.mod
Comment (priteshbandi) [reply go.mod]: Yeah, I am planning to release `notation-plugin-framework-go` once this PR is ready to be merged(wont merge it until we release `notation-plugin-framework-go`.

Comment (shizhMSFT) [reply go.mod]: They are `golang.org/x` packages and I think they will never be stabilized. The reason is simple that once they are stabilized, they will be moved to the golang built-in libraries.

> These repositories are part of the Go Project but outside the main Go tree. They are developed under looser compatibility requirements than the Go core.

Comment (shizhMSFT) [reply go.mod]: Basically, since `notation-go` depends on `notation-plugin-framework-go` and expose `notation-plugin-framework-go` as part of its API. The instability of `notation-plugin-framework-go` makes `notation-go` instable. However, `notation-go` does not expose `golang.org/x` packages as part of its API.

Comment (shizhMSFT) [reply go.mod]: > Basically, since `notation-go` depends on `notation-plugin-framework-go` and expose `notation-plugin-framework-go` as part of its API.

I find a solution. How about we keep everything under https://github.com/notaryproject/notation-go/tree/main/plugin and mark all of them `Deprecated`. Meanwhile, we replace the implementation of `signer/plugin.go` and `verifier/verifier.go` with `notation-plugin-framework-go` as you've done in this PR?

Comment (priteshbandi) [reply go.mod]: umm I think we are not going to make any breaking changes in `notation-plugin-framework-go` so after next planned release (this week?) we can take dependency on it and before before next to next release, we will release `notation-plugin-framework-go-1.0`. Does that sounds fine? Also, we have enough unit/tests to avoid any breaking change in notation-go.

PS: I wont merge this PR until next release,

Comment (priteshbandi) [reply go.mod]: Updated `go.mod` to depend upon notation-plugin-framework-go-1.0

File: plugin/proto/proto.go
Comment (JeyJeyGao) [inline plugin/proto/proto.go:21]: Will we replace it with `notation-plugin-framework-go/plugin`

File: plugin/proto/proto.go
Comment (JeyJeyGao) [inline plugin/proto/proto.go:24]: Will we replace it with `notation-plugin-framework-go/plugin`

File: plugin/plugin.go
Comment (JeyJeyGao) [inline plugin/plugin.go:110]: Will we replace `proto.ContractVersion` with `plugin.ContractVersion` in `notation-plugin-framework-go/plugin`?

**Review by JeyJeyGao — COMMENTED** [review]
There are still many instances where the code uses `proto.xxx` that have not been replaced with `notation-plugin-framework-go/plugin`. Will we replace them?

File: plugin/proto/proto.go
Comment (priteshbandi) [reply plugin/proto/proto.go]: Done

File: plugin/plugin.go
Comment (priteshbandi) [reply plugin/plugin.go]: Done

**Comment by priteshbandi** [comment]
> There are still many instances where the code uses `proto.xxx` that have not been replaced with `notation-plugin-framework-go/plugin`. Will we replace them?

Some code such as supporting methods not present `notation-plugin-framework-go`, unit tests will continue to use `proto.xxx`. We are not updating unit test to make sure we are backward compatible.

File: plugin/proto/algorithm_test.go
Comment (JeyJeyGao) [inline plugin/proto/algorithm_test.go:100]: should we keep `HashAlgorithm` for unit test, instead of `plugin.HashAlgorithm`?

File: signer/plugin_test.go
Comment (JeyJeyGao) [inline signer/plugin_test.go:118]: Is it a little **breaking change**? We may need to mention it in the release.

File: plugin/integration_test.go
Comment (JeyJeyGao) [inline plugin/integration_test.go:94]: Should we keep `proto.GetMetadataRequest{}` for unit test

File: signer/plugin_test.go
Comment (priteshbandi) [reply signer/plugin_test.go:118]: Yes looks like it is a small breaking change.

File: plugin/integration_test.go
Comment (priteshbandi) [reply plugin/integration_test.go:94]: this is integration test. We have kept proto for unit tests.

**Review by Two-Hearts — APPROVED** [review]
LGTM
I think we should wait for Shiwei's comments regarding the PR.

**Comment by JeyJeyGao** [comment]
LGTM 
Pending on @shizhMSFT

File: plugin/proto/algorithm.go
Comment (shizhMSFT) [inline plugin/proto/algorithm.go:38]: Should those consts also be marked as deprecated?

File: plugin/proto/algorithm.go
Comment (shizhMSFT) [inline plugin/proto/algorithm.go:105]: Same here

File: plugin/proto/algorithm.go
Comment (shizhMSFT) [inline plugin/proto/algorithm.go:147]: Same here

File: plugin/proto/errors.go
Comment (shizhMSFT) [inline plugin/proto/errors.go:49]: Same here

File: plugin/proto/proto.go
Comment (shizhMSFT) [inline plugin/proto/proto.go:31]: Same here

File: plugin/proto/proto.go
Comment (shizhMSFT) [inline plugin/proto/proto.go:63]: same here

File: plugin/proto/proto.go
Comment (shizhMSFT) [inline plugin/proto/proto.go:86]: same here

**Review by shizhMSFT — APPROVED** [review]
LGTM with suggestions on deprecation notices

**Comment by priteshbandi** [comment]
> LGTM with suggestions on deprecation notices

Will merge the change and raised new PR for your suggestions.

File: plugin/proto/algorithm.go
Comment (priteshbandi) [reply plugin/proto/algorithm.go:38]: addressed in https://github.com/notaryproject/notation-go/pull/382

File: plugin/proto/algorithm.go
Comment (priteshbandi) [reply plugin/proto/algorithm.go:147]: addressed in https://github.com/notaryproject/notation-go/pull/382

File: plugin/proto/proto.go
Comment (priteshbandi) [reply plugin/proto/proto.go:31]: addressed in https://github.com/notaryproject/notation-go/pull/382

File: plugin/proto/proto.go
Comment (priteshbandi) [reply plugin/proto/proto.go:63]: addressed in https://github.com/notaryproject/notation-go/pull/382

File: plugin/proto/errors.go
Comment (priteshbandi) [reply plugin/proto/errors.go:49]: addressed in https://github.com/notaryproject/notation-go/pull/382

File: plugin/proto/proto.go
Comment (priteshbandi) [reply plugin/proto/proto.go:63]: addressed in https://github.com/notaryproject/notation-go/pull/382

File: plugin/proto/algorithm.go
Comment (priteshbandi) [reply plugin/proto/algorithm.go:105]: addressed in https://github.com/notaryproject/notation-go/pull/382

**Review by rgnote — APPROVED** [review]
LGTM

**Pull request merged**
